plugins/builds/update.js

```diff
@@ -245,16 +270,99 @@ module.exports = () => ({
+                        { pipeline, job, build: newBuild, username, scmContext, event: newEvent, stage },
+                        request.server.app
+                    );
                 }
+                // Do not continue downstream is current job is stage teardown and statusBuild has failure
+            } else if (newBuild.status === 'SUCCESS' && isStageTeardown && stageBuildHasFailure) {
+                await removeJoinBuilds(
+                    { pipeline, job, build: newBuild, username, scmContext, event: newEvent, stage },
+                    request.server.app
+                );
             } else {
-                await triggerNextJobs({ pipeline, job, build: newBuild, username, scmContext }, request.server.app);
+                await triggerNextJobs(
+                    { pipeline, job, build: newBuild, username, scmContext, event: newEvent, stage },
+                    request.server.app
+                );
+            }
+
+            // Determine if stage teardown build should start
+            // (if stage teardown build exists, and stageBuild.status is negative,
+            // and there are no active stage builds, and teardown build is not started)
+            if (stage) {
+                const stageTeardownName = getFullStageJobName({ stageName: stage.name, jobName: 'teardown' });
+                const stageTeardownJob = await jobFactory.get({
+                    params: { pipelineId: pipeline.id, name: stageTeardownName }
+                });
+                let stageTeardownBuild;
+
+                if (stageTeardownJob) {
+                    // Get stage teardown build
+                    stageTeardownBuild = await buildFactory.get({
+                        params: { eventId: newEvent.id, jobId: stageTeardownJob.id }
+                    });
+
+                    // Get all jobIds for jobs in the stage
+                    const stageJobIds = stage.jobIds;
+
+                    stageJobIds.push(stage.setup);
+
+                    // Get all builds in a stage
+                    const stageJobBuilds = await buildFactory.list({
+                        params: { jobId: stageJobIds, eventId: newEvent.id }
+                    });
+                    let stageIsDone = false;
+
+                    if (stageJobBuilds && stageJobBuilds.length !== 0) {
+                        stageIsDone = !stageJobBuilds.some(b => !FINISHED_STATUSES.includes(b.status));
+                    }
+
+                    // Start stage teardown build
+                    if (
+                        stageTeardownBuild &&
+                        stageTeardownBuild.status === 'CREATED' &&
+                        stageBuildHasFailure &&
+                        stageIsDone
+                    ) {
+                        stageTeardownBuild.status = 'QUEUED';
+                        await stageTeardownBuild.update();
+                        await stageTeardownBuild.start();
+                    }
```

Comment:
If the `teardown` build is already started, we don't need to query for other stage job builds and determine if the stage is done.

Also move the logic to determine whether the stage is done to. a separate function.
```suggestion
if(stageTeardownBuild.status === 'CREATED' && isStageDone(stage)) {
    stageTeardownBuild.status = 'QUEUED';
    await stageTeardownBuild.update();
    await stageTeardownBuild.start();
}
```


```
isStageDone(stage){
// Get all jobIds for jobs in the stage
                    const stageJobIds = stage.jobIds;

                    stageJobIds.push(stage.setup);

                    // Get all builds in a stage
                    const stageJobBuilds = await buildFactory.list({
                        params: { jobId: stageJobIds, eventId: newEvent.id }
                    });
                    let stageIsDone = false;

                    if (stageJobBuilds && stageJobBuilds.length !== 0) {
                        stageIsDone = !stageJobBuilds.some(b => !FINISHED_STATUSES.includes(b.status));
                    }

return stageIsDone;
}
```